java/ql/src/experimental/Security/CWE/CWE-749/UnsafeAndroidAccess.ql

```diff
@@ -0,0 +1,139 @@
+/**
+ * @name Unsafe resource loading in Android webview
+ * @description JavaScript rendered inside WebViews can access any protected application file and web resource from any origin
+ * @kind path-problem
+ * @tags security
+ *       external/cwe/cwe-749
+ *       external/cwe/cwe-079
+ */
+
+import java
+import semmle.code.java.frameworks.android.Intent
+import semmle.code.java.frameworks.android.WebView
+import semmle.code.java.dataflow.FlowSources
+
+/**
+ * Allow universal access methods in the WebSettings class
+ */
+class AllowUniversalAccessMethod extends Method {
+  AllowUniversalAccessMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    (
+      this.hasName("setAllowUniversalAccessFromFileURLs") or
+      this.hasName("setAllowFileAccessFromFileURLs")
+    )
+  }
+}
+
+/**
+ * `setJavaScriptEnabled` method for the webview
+ */
+class AllowJavaScriptMethod extends Method {
+  AllowJavaScriptMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    this.hasName("setJavaScriptEnabled")
+  }
+}
+
+/**
+ * Check whether JavaScript is enabled in the webview with universal resource access
+ */
+predicate isJSEnabled(MethodAccess ma) {
+  exists(VarAccess va, MethodAccess jsa |
+    ma.getQualifier() = va and
+    jsa.getQualifier() = va.getVariable().getAnAccess() and
+    jsa.getMethod() instanceof AllowJavaScriptMethod and
+    jsa.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+  )
+}
+
+/**
+ * Load URL method call on the `android.webkit.WebView` object
+ */
+class LoadResourceMethodAccess extends MethodAccess {
```

Comment:
Load -> Fetch everywhere, since it does both the load and post methods

Reply:
Updated